lib/r10k/action/deploy/environment.rb

```diff
@@ -43,6 +45,50 @@ def call
 
         private
 
+        def extract_credentials!
+          if @sshkey_path && @token_path
+            raise R10K::Error, "Cannot specify both an SSH key and a token to use with this deploy."
+          end
+
+          if @sshkey_path
+            if @sshkey_path == '-'
+              sshkey = $stdin.read
+              keypath = write_key_to_tmpfile(sshkey).path
+              { sshkey_path: keypath }
+            elsif File.exists?(@sshkey_path)
+              { sshkey_path: @sshkey_path }
+            else
+              raise R10K::Error, _("{%path} does not exist, cannot load SSH key") % { path: @sshkey_path }
+            end
+          elsif @token_path
+            if @token_path == '-'
+              token = $stdin.read
+            elsif File.exists?(@token_path)
+              token = File.read(@token_path).strip
+            else
+              raise R10K::Error, _("{%path} does not exist, cannot load OAuth token") % { path: @token_path }
+            end
+
+            unless valid_token?(token)
+              raise R10K::Error, _("Supplied token contains invalid characters.")
+            end
+
+            { token: token }
+          end
+        end
+
+        def write_key_to_tmpfile(sshkey)
+          keypath = Tempfile.new('sshkey')
+          File.chmod(0600, keypath)
+          keypath.write(sshkey)
+          keypath.close
+          keypath
+        end
```

Comment:
I am vaguely worried about leaving behind keys in the Tmpfile dir, and there are some recommendations from ruby-doc about making this [more secure](https://ruby-doc.org/stdlib-2.4.0/libdoc/tempfile/rdoc/Tempfile.html#method-i-unlink-label-Unlink-before-close). 

Posting the relevant quote here:

> On POSIX systems, it’s possible to unlink a file right after creating it, and before closing it. This removes the filesystem entry without closing the file handle, so it ensures that only the processes that already had the file handle open can access the file’s contents. It’s strongly recommended that you do this if you do not want any other processes to be able to read from or write to the Tempfile, and you do not need to know the Tempfile’s filename either.

I guess the catch is Windows isn't super great about this either...but, we don't have PE on windows...yet...

Reply:
I think we do need to know the name though. The thing Rugged saves and loads is a file name...

I think the ultimate solution here is to use a regular file that is encrypted with a password and has locked down perms, and then clean it up by hand when we know we're done.

Reply:
I'm not sure if there's an interim thing we could be doing that might be more secure though. Thoughts, @justinstoller @pcarlisle ?

Reply:
The file should be created in a directory that is locked down and owned by the process (e.g. `~/.ssh`). I assume somewhere in `/opt/puppetlabs` will work for PE and mostlikely is the process user's homedir (if it exists) and/or the configured datadir.

I'm not sure about unlinking before closing, we will be the same process as rugged right? But not if we provide support for shellgit at some point?

As a real handwaving thing, I would look to what `puppet` uses as that code base has learned (the hard way) how to secure tmp files. [FileSystem impl](https://github.com/puppetlabs/puppet/blob/main/lib/puppet/file_system/file_impl.rb#L168-L183) & [FileSystem::UniqueFile](https://github.com/puppetlabs/puppet/blob/cd7e16542c919e516dda44dbf6bac0ff64dacf0a/lib/puppet/file_system/uniquefile.rb#L10).

Reply:
Also.... If we're writing files now, we can do that in code-manager where we know the datadir and just give the path to r10k. Right?

Reply:
Yeah, that will work for our usecase, but we probably want to make sure we're still doing something sensible for FOSS users, I think? Does `~/.ssh` seem like the right call for the general case?

Reply:
Or wait. Are you saying we should just only accept a filepath (and not stdin) for this? Cause that works for me too.

Reply:
I'm a little reticent to put items in `.ssh` that are ephemeral by nature and could get left behind. Do we need our own cool `.r10k-ssh` dir? 🤷

Reply:
I'm also totally fine changing this to only accept a file path (since that's what we need anyway) and not messing with this at all.

Reply:
I'm thinking, the "-" stdin bit was good when our explicit goal was to not write anything to disk.

But, iiuc, we've landed on writing to disk is unavoidable.

Then for our use case we should write the secrets to disk where it makes the most sense for us. When writing to secrets to disk we should write them to a secure directory owned by the process (a directory locked down like ~/.ssh normally is). For us, I think that is our datadir in code-manager (and if not we can make a sub-directory and manage its perms).

Figuring out that where-to-write-secrets bit sounds like it would be hard from r10k (haven't done a lot of looking though, maybe I'm wrong?) and seems like we might as well write the secret to disk in the application calling r10k since it is explicitly configured with the process owned datadir.

Reply:
That works for me. I'll update this. I'll leave `-` as an option for the token though, since there we really don't ever want to write it out.